File: booking-platform/backend/app/services/show_service.py
from sqlalchemy.orm import Session

from app.models.show import Show
from app.repositories import movie_repository,show_repository
from app.schemas.show import ShowCreate
from app.services.movie_service import MovieNotFoundError

def get_shows_for_movie(db:Session,movie_id:int) -> list[Show]:
    """ 
    Return all shows for a movie.
    Validates the movie exists first - no point querying shows for a ghost movie.
    Raises MovieNotFoundError if the movie doesn't exist (route converts to 404).

    """
    
    if movie_repository.get_by_id(db,movie_id) is None:
        raise ModuleNotFoundError(f"Movie with id {movie_id} not found")
    return show_repository.get_by_movie(db,movie_id)


def create_show(db:Session,payload:ShowCreate) -> Show:
    """
    Create a new show for a movie.

    Business rules enforced here:
        1. The movie must exist - we validate the movie_id before inserting.
        2. available seats is intialized to the total_seats - the client never sets this.
           It will be decremented automatically when the bookings are made.

    Raises MovieNotFoundError if the movie_id in the payload doesn't exist.

    """

    if movie_repository.get_by_id(db,payload.movie_id) is None:
        raise ModuleNotFoundError(f"Movie with id {payload.movie_id} not found")
    
    data = payload.model_dump()

    # Enforces the business rule : new shows start fully available.
    # The client sends total_seats - we derive available_seats from it.

    data["available_seats"] = data["total_seats"]

    return show_repository.create(db,data)

File: booking-platform/frontend/src/features/shows/services/showService.ts
/*
 * showService.ts — Shows API Layer
 */

import apiClient from "../../../services/apiClient";
import { ENDPOINTS } from "../../../services/endpoints";
import type { ShowResponse } from "../../movies/types";

// Type for creating a new show
export interface ShowCreate {
  movie_id: number;
  show_time: string; // ISO datetime: "2024-06-15T19:30:00"
  total_seats: number;
  hall_name: string;
}

// Get all shows for a movie
export async function getShowsForMovie(
  movieId: number,
): Promise<ShowResponse[]> {
  const response = await apiClient.get<ShowResponse[]>(
    ENDPOINTS.movies.shows(movieId),
  );
  return response.data;
}

// Create a new show
export async function createShow(data: ShowCreate): Promise<ShowResponse> {
  const response = await apiClient.post<ShowResponse>(
    ENDPOINTS.shows.create,
    data,
  );
  return response.data;
}
